File: services/worker/src/job_getter_worker/runner/adapters/greenhouse.py
"""The Greenhouse adapter, version 1.

Greenhouse is the spec's first target ("Target Greenhouse first, Lever
second"). This adapter reads a Greenhouse-shaped application form: the
``#application_form`` / ``#application-form`` container boards have used for
years, with its labelled inputs, its ``required`` markers and its file input
for the CV.

**What "tested" means here, exactly.** The snippets below are exercised against
a synthetic Greenhouse-shaped page in
``fixtures/ats-pages/greenhouse-application.html``, driven by a real Chromium. They
have not been run against a live Greenhouse board, and no application has been
submitted through them. The support matrix must say that rather than "Greenhouse
supported"; "support badges must reflect actual tested adapters"
(docs/spec/12_IMPLEMENTATION_PLAN.md), and a fixture is not a live board.

The adapter claims a page by host *or* by that container's presence. Structure
matters as much as host because Greenhouse serves the same form from
``boards.greenhouse.io``, ``job-boards.greenhouse.io`` and from employers' own
domains through an embed - and because an adapter that only recognises a
hostname is one redirect away from filling a page it has never seen.
"""

from __future__ import annotations

from typing import Any, Final
from urllib.parse import urlsplit

from ..forms import FormSchema
from .base import parse_fields

#: Hosts Greenhouse itself serves application forms from.
GREENHOUSE_HOSTS: Final = (
    "boards.greenhouse.io",
    "job-boards.greenhouse.io",
    "my.greenhouse.io",
)

#: The container every Greenhouse application form has had. Also what the
#: synthetic fixture carries, because the fixture is a copy of the shape.
FORM_SELECTOR: Final = "#application_form, #application-form, form[action*='greenhouse']"

_MARKER_SCRIPT: Final = f"""
() => document.querySelector({FORM_SELECTOR!r}) !== null
"""

_IDENTITY_SCRIPT: Final = """
() => {
  const text = (selector) => {
    const node = document.querySelector(selector);
    return node ? node.textContent.trim() : '';
  };
  return {
    company:
      text('[data-qa="company-name"]') ||
      text('.company-name') ||
      (document.querySelector('meta[property="og:site_name"]')?.content ?? ''),
    title: text('.app-title') || text('h1') || document.title,
  };
}
"""

#: Reads the form once and returns a plain description. Everything
#: interpretive - what counts as required, what a question is called, which
#: widgets are unsupported - happens in Python, where it is unit-tested.
_INSPECT_SCRIPT: Final = f"""
() => {{
  const form = document.querySelector({FORM_SELECTOR!r});
  if (!form) return [];

  const labelFor = (el) => {{
    if (el.getAttribute('aria-label')) return el.getAttribute('aria-label');
    const labelledBy = el.getAttribute('aria-labelledby');
    if (labelledBy) {{
      const parts = labelledBy.split(/\\s+/)
        .map((id) => document.getElementById(id))
        .filter(Boolean)
        .map((node) => node.textContent);
      if (parts.length) return parts.join(' ');
    }}
    if (el.id) {{
      const explicit = form.querySelector(`label[for="${{CSS.escape(el.id)}}"]`);
      if (explicit) return explicit.textContent;
    }}
    const wrapping = el.closest('label');
    if (wrapping) return wrapping.textContent;
    const group = el.closest('fieldset');
    const legend = group ? group.querySelector('legend') : null;
    if (legend) return legend.textContent;
    return el.getAttribute('placeholder') || el.name || '';
  }};

  const clean = (value) => (value || '').replace(/\\s+/g, ' ').replace(/\\*$/, '').trim();
  const isRequired = (el, label) =>
    el.required === true ||
    el.getAttribute('aria-required') === 'true' ||
    /\\*\\s*$/.test(label) ||
    /\\(required\\)/i.test(label);

  const selectorFor = (el) => {{
    if (el.id) return `#${{CSS.escape(el.id)}}`;
    if (el.name) return `${{el.tagName.toLowerCase()}}[name="${{CSS.escape(el.name)}}"]`;
    return '';
  }};

  const out = [];
  const seenGroups = new Set();

  for (const el of form.querySelectorAll('input, select, textarea, [role="combobox"]')) {{
    if (el.disabled) continue;
    const type = (el.getAttribute('type') || '').toLowerCase();
    if (type === 'hidden' || type === 'submit' || type === 'button') continue;

    const rawLabel = labelFor(el);
    const label = clean(rawLabel);
    const required = isRequired(el, rawLabel);

    if (el.tagName === 'SELECT') {{
      out.push({{
        selector: selectorFor(el),
        label,
        kind: 'select',
        required,
        options: Array.from(el.options)
          .filter((option) => option.value !== '')
          .map((option) => clean(option.textContent)),
        option_values: Array.from(el.options)
          .filter((option) => option.value !== '')
          .map((option) => option.value),
      }});
      continue;
    }}

    if (el.tagName === 'TEXTAREA') {{
      out.push({{ selector: selectorFor(el), label, kind: 'textarea', required, options: [] }});
      continue;
    }}

    if (type === 'radio' || type === 'checkbox') {{
      // One entry per group, named by its legend, with every offered option.
      const groupKey = `${{type}}:${{el.name}}`;
      if (seenGroups.has(groupKey)) continue;
      seenGroups.add(groupKey);
      const members = Array.from(form.querySelectorAll(
        `input[type="${{type}}"][name="${{CSS.escape(el.name)}}"]`,
      ));
      const fieldset = el.closest('fieldset');
      const legend = fieldset ? clean(fieldset.querySelector('legend')?.textContent) : '';
      out.push({{
        selector: `input[type="${{type}}"][name="${{CSS.escape(el.name)}}"]`,
        label: legend || label,
        kind: type,
        required: members.some((member) => isRequired(member, labelFor(member))) ||
          (fieldset ? /\\*\\s*$/.test(legend) : false),
        options: members.map((member) => clean(labelFor(member)) || member.value),
        option_values: members.map((member) => member.value),
      }});
      continue;
    }}

    if (el.getAttribute('role') === 'combobox' || el.tagName !== 'INPUT') {{
      // A custom widget: recognised, named, and reported as undriveable.
      out.push({{ selector: selectorFor(el), label, kind: 'custom', required, options: [] }});
      continue;
    }}

    out.push({{
      selector: selectorFor(el),
      label,
      kind: type || 'text',
      required,
      options: [],
    }});
  }}
  return out;
}}
"""


#: Read a post-submission confirmation, if the page is showing one.
#:
#: Deliberately conservative. It looks for Greenhouse's own confirmation
#: container and for a small set of phrasings the board actually uses, and it
#: returns ``null`` for anything else. The cost of the two mistakes is not
#: symmetric: failing to recognise a real confirmation leaves the user to say
#: so themselves, while inventing one from a stray "thank you" on a marketing
#: footer would record a submission that may never have happened.
_CONFIRMATION_SCRIPT: Final = r"""
() => {
  const clean = (value) => (value || '').replace(/\s+/g, ' ').trim();

  // Greenhouse's own confirmation container, when the board renders one.
  const container = document.querySelector(
    '#application_confirmation, .application-confirmation, [data-qa="confirmation"]'
  );

  let text = container ? clean(container.textContent) : '';

  if (!text) {
    // Otherwise accept only an explicit statement, and only from a heading or
    // a status region -- not from anywhere on the page.
    const candidates = document.querySelectorAll(
      'h1, h2, h3, [role="status"], [role="alert"], .flash-message, .status-message'
    );
    const accepted = new RegExp(
      '(application (was )?(successfully )?(submitted|received)' +
        '|thank you for applying' +
        '|your application has been (submitted|received))',
      'i'
    );
    for (const node of candidates) {
      const value = clean(node.textContent);
      if (value && accepted.test(value)) {
        text = value;
        break;
      }
    }
  }

  if (!text) return null;

  // An application/reference number when the page shows one beside the text.
  const body = clean(document.body ? document.body.textContent : '');
  const match = body.match(
    new RegExp(
      '(?:application|reference|confirmation)\\s*(?:id|number|no\\.?|#)' +
        '\\s*[:#]?\\s*([A-Za-z0-9-]{4,40})',
      'i'
    )
  );

  return {
    confirmation_text: text.slice(0, 2000),
    reference: match ? match[1] : null,
    url: window.location.href,
  };
}
"""


class GreenhouseAdapter:
    """Reads Greenhouse-shaped application forms."""

    name = "greenhouse"
    version = "greenhouse/v1"

    def handles(self, url: str, marker_found: bool) -> bool:
        host = urlsplit(url).hostname or ""
        if host in GREENHOUSE_HOSTS or host.endswith(".greenhouse.io"):
            return True
        # An embed on the employer's own domain is still a Greenhouse form, and
        # the container is the thing this adapter actually knows how to drive.
        return marker_found

    @property
    def marker_script(self) -> str:
        return _MARKER_SCRIPT

    @property
    def identity_script(self) -> str:
        return _IDENTITY_SCRIPT

    @property
    def inspect_script(self) -> str:
        return _INSPECT_SCRIPT

    def parse_inspection(self, raw: Any) -> FormSchema:
        return FormSchema(fields=parse_fields(raw))

    @property
    def confirmation_script(self) -> str:
        return _CONFIRMATION_SCRIPT

File: apps/extension/src/adapters/greenhouse.ts
/**
 * Reading a Greenhouse-shaped application form from the DOM.
 *
 * A port of the JavaScript snippets in
 * `services/worker/src/job_getter_worker/runner/adapters/greenhouse.py`, which
 * the Python runner evaluates through Playwright. They were already plain
 * DOM JavaScript with no Playwright API inside them, so this is the same code
 * in TypeScript rather than a second reader — and it returns the same plain
 * shape, which `@job-getter/fill-planner` then interprets identically for both
 * clients.
 *
 * **What "tested" means here, exactly.** These are exercised against the same
 * synthetic page the Python adapter is, `fixtures/ats-pages/greenhouse-application.html`,
 * through jsdom. They have not been run against a live Greenhouse board and no
 * application has been submitted through them. The support matrix must say
 * that rather than "Greenhouse supported": a fixture is not a live board.
 *
 * Nothing here interprets. It does not decide what a question is called, which
 * widget is unsupported or what may be typed — it reports what it saw, and the
 * shared planner decides. That split is why the interesting behaviour is
 * testable without a browser at all.
 */

export const ADAPTER_NAME = 'greenhouse';
export const ADAPTER_VERSION = 'v1';

/** Hosts Greenhouse itself serves application forms from. */
export const GREENHOUSE_HOSTS: readonly string[] = [
  'boards.greenhouse.io',
  'job-boards.greenhouse.io',
  'my.greenhouse.io',
];

/**
 * The container every Greenhouse application form has had. Also what the
 * synthetic fixture carries, because the fixture is a copy of the shape.
 */
export const FORM_SELECTOR = "#application_form, #application-form, form[action*='greenhouse']";

/**
 * Whether this adapter claims the page.
 *
 * Structure matters as much as host, because Greenhouse serves the same form
 * from several of its own domains and from employers' own domains through an
 * embed — and because an adapter that only recognises a hostname is one
 * redirect away from filling a page it has never seen.
 */
export function handles(url: string, markerFound: boolean): boolean {
  if (markerFound) return true;
  try {
    return GREENHOUSE_HOSTS.includes(new URL(url).hostname.toLowerCase());
  } catch {
    return false;
  }
}

export function findForm(root: Document): Element | null {
  return root.querySelector(FORM_SELECTOR);
}

function clean(value: string | null | undefined): string {
  return (value ?? '').replace(/\s+/g, ' ').replace(/\*$/, '').trim();
}

export function readIdentity(root: Document): { company: string; title: string } {
  const text = (selector: string): string => {
    const node = root.querySelector(selector);
    return node?.textContent?.trim() ?? '';
  };
  const siteName = root.querySelector<HTMLMetaElement>('meta[property="og:site_name"]')?.content;
  return {
    company: text('[data-qa="company-name"]') || text('.company-name') || (siteName ?? ''),
    title: text('.app-title') || text('h1') || root.title,
  };
}

/** One row of the shape `@job-getter/fill-planner`'s `parseFields` consumes. */
export interface RawFieldRow {
  selector: string;
  label: string;
  kind: string;
  required: boolean;
  options: string[];
  option_values?: string[];
}

function labelFor(form: Element, root: Document, el: Element): string {
  const aria = el.getAttribute('aria-label');
  if (aria) return aria;

  const labelledBy = el.getAttribute('aria-labelledby');
  if (labelledBy) {
    const parts = labelledBy
      .split(/\s+/)
      .map((id) => root.getElementById(id))
      .filter((node): node is HTMLElement => node !== null)
      .map((node) => node.textContent ?? '');
    if (parts.length > 0) return parts.join(' ');
  }

  if (el.id) {
    const explicit = form.querySelector(`label[for="${cssEscape(el.id)}"]`);
    if (explicit) return explicit.textContent ?? '';
  }

  const wrapping = el.closest('label');
  if (wrapping) return wrapping.textContent ?? '';

  const group = el.closest('fieldset');
  const legend = group?.querySelector('legend');
  if (legend) return legend.textContent ?? '';

  return el.getAttribute('placeholder') || (el as HTMLInputElement).name || '';
}

function isRequired(el: Element, label: string): boolean {
  return (
    (el as HTMLInputElement).required === true ||
    el.getAttribute('aria-required') === 'true' ||
    /\*\s*$/.test(label) ||
    /\(required\)/i.test(label)
  );
}

function selectorFor(el: Element): string {
  if (el.id) return `#${cssEscape(el.id)}`;
  const name = (el as HTMLInputElement).name;
  if (name) return `${el.tagName.toLowerCase()}[name="${cssEscape(name)}"]`;
  return '';
}

/** `CSS.escape` is absent from some test environments; the fallback is exact. */
function cssEscape(value: string): string {
  const escaper = (globalThis as { CSS?: { escape?: (input: string) => string } }).CSS?.escape;
  if (escaper) return escaper(value);
  return value.replace(/[^\w-]/g, (character) => `\\${character}`);
}

/**
 * Read the form once and return a plain description.
 *
 * Deliberately a faithful port, including the order of the branches: a
 * `<select>`, then a `<textarea>`, then a radio or checkbox *group* (one entry
 * per group, named by its legend, carrying every offered option), then
 * anything with `role="combobox"` or any non-`<input>` element as an
 * explicitly unsupported widget, then a plain input.
 */
export function readFields(root: Document): RawFieldRow[] {
  const form = findForm(root);
  if (form === null) return [];

  const out: RawFieldRow[] = [];
  const seenGroups = new Set<string>();

  for (const el of form.querySelectorAll('input, select, textarea, [role="combobox"]')) {
    if ((el as HTMLInputElement).disabled) continue;
    const type = (el.getAttribute('type') ?? '').toLowerCase();
    if (type === 'hidden' || type === 'submit' || type === 'button') continue;

    const rawLabel = labelFor(form, root, el);
    const label = clean(rawLabel);
    const required = isRequired(el, rawLabel);

    if (el.tagName === 'SELECT') {
      const options = Array.from((el as HTMLSelectElement).options).filter(
        (option) => option.value !== '',
      );
      out.push({
        selector: selectorFor(el),
        label,
        kind: 'select',
        required,
        options: options.map((option) => clean(option.textContent)),
        option_values: options.map((option) => option.value),
      });
      continue;
    }

    if (el.tagName === 'TEXTAREA') {
      out.push({ selector: selectorFor(el), label, kind: 'textarea', required, options: [] });
      continue;
    }

    if (type === 'radio' || type === 'checkbox') {
      const name = (el as HTMLInputElement).name;
      const groupKey = `${type}:${name}`;
      if (seenGroups.has(groupKey)) continue;
      seenGroups.add(groupKey);

      const members = Array.from(
        form.querySelectorAll<HTMLInputElement>(`input[type="${type}"][name="${cssEscape(name)}"]`),
      );
      const fieldset = el.closest('fieldset');
      const legend = fieldset ? clean(fieldset.querySelector('legend')?.textContent) : '';
      out.push({
        selector: `input[type="${type}"][name="${cssEscape(name)}"]`,
        label: legend || label,
        kind: type,
        required:
          members.some((member) => isRequired(member, labelFor(form, root, member))) ||
          (fieldset ? /\*\s*$/.test(legend) : false),
        options: members.map((member) => clean(labelFor(form, root, member)) || member.value),
        option_values: members.map((member) => member.value),
      });
      continue;
    }

    if (el.getAttribute('role') === 'combobox' || el.tagName !== 'INPUT') {
      // A custom widget: recognised, named, and reported as undriveable.
      out.push({ selector: selectorFor(el), label, kind: 'custom', required, options: [] });
      continue;
    }

    out.push({ selector: selectorFor(el), label, kind: type || 'text', required, options: [] });
  }

  return out;
}

/** What a confirmation page said, in the shape the runner's script returns. */
export interface RawConfirmation {
  readonly confirmation_text: string;
  readonly reference: string | null;
  readonly url: string;
}

/**
 * Whitespace only. Not `clean`, which also drops a trailing required-marker
 * asterisk from labels: the runner's confirmation script does not, and a
 * reader that disagreed with it about one character would store different
 * evidence for the same page.
 */
function collapse(value: string | null | undefined): string {
  return (value ?? '').replace(/\s+/g, ' ').trim();
}

const CONFIRMATION_CONTAINER =
  '#application_confirmation, .application-confirmation, [data-qa="confirmation"]';
const CONFIRMATION_CANDIDATES =
  'h1, h2, h3, [role="status"], [role="alert"], .flash-message, .status-message';
const ACCEPTED_CONFIRMATION = new RegExp(
  '(application (was )?(successfully )?(submitted|received)' +
    '|thank you for applying' +
    '|your application has been (submitted|received))',
  'i',
);
// A literal, not `new RegExp('…')`: in a string, `\s` has to be written `\\s`,
// and a lost backslash silently turns "whitespace" into the letter s.
const CONFIRMATION_REFERENCE =
  /(?:application|reference|confirmation)\s*(?:id|number|no\.?|#)\s*[:#]?\s*([A-Za-z0-9-]{4,40})/i;

/**
 * Read a post-submission confirmation, if the page is showing one.
 *
 * The runner's `_CONFIRMATION_SCRIPT`, line for line, and pinned to it by
 * `fixtures/fill-planner/greenhouse-confirmation.json`. Deliberately
 * conservative, for the same reason: failing to recognise a real
 * confirmation leaves the person to say so themselves, while inventing one
 * from a stray "thank you" in a footer would record a submission that may
 * never have happened.
 */
export function readConfirmation(root: Document, url: string): RawConfirmation | null {
  const container = root.querySelector(CONFIRMATION_CONTAINER);
  let text = container ? collapse(container.textContent) : '';

  if (!text) {
    // Otherwise accept only an explicit statement, and only from a heading or
    // a status region, not from anywhere on the page.
    for (const node of Array.from(root.querySelectorAll(CONFIRMATION_CANDIDATES))) {
      const value = collapse(node.textContent);
      if (value && ACCEPTED_CONFIRMATION.test(value)) {
        text = value;
        break;
      }
    }
  }

  if (!text) return null;

  // An application or reference number when the page shows one beside the text.
  const body = collapse(root.body ? root.body.textContent : '');
  const match = body.match(CONFIRMATION_REFERENCE);
  return {
    confirmation_text: text.slice(0, 2000),
    reference: match ? (match[1] ?? null) : null,
    url,
  };
}
